Examples/inference/mpt README updates

sakshigupta93 opened the pull request:
Updated just one README for now + added high level README. If everything looks good I'll add a similar README for instructor and diffusion model repos as well. 

Main README
<img width="876" alt="Screenshot 2023-06-12 at 3 44 51 PM" src="https://github.com/mosaicml/examples/assets/1689437/f49c7d8a-bfc6-4b03-b389-f4f0e7fc574f">


MPT README
<img width="908" alt="Screenshot 2023-06-12 at 3 45 20 PM" src="https://github.com/mosaicml/examples/assets/1689437/2e6bf2f3-c87f-42de-a085-efc35ceb24a7">
<img width="908" alt="Screenshot 2023-06-12 at 3 45 29 PM" src="https://github.com/mosaicml/examples/assets/1689437/f9c830dd-e931-47b3-9165-c304270ba364">
<img width="971" alt="Screenshot 2023-06-12 at 3 45 35 PM" src="https://github.com/mosaicml/examples/assets/1689437/bdb20136-4fcf-44cf-a3aa-af0c013c3cb3">

dakinggg commented on examples/inference-deployments/mpt/README.md:
```diff
@@ -1,22 +1,57 @@
-## Model Description
+## Inference with MPT-7B 
```
Can we please include a "setup" section that tells the user how to setup their environment and account to actually be able to run this?

dakinggg replied:
this could also be a pointer to a top level readme with these instructions

sakshigupta93 replied:
Sounds good, I did add it at the top level draft that I have. Will edit!

dakinggg commented on examples/inference-deployments/mpt/README.md:
```diff
@@ -1,22 +1,57 @@
+## Inference with MPT-7B 
 
-A state-of-the-art 6.7B parameter instruction finetuned language model trained by MosaicML. Check out the [MPT-Instruct huggingface repo](https://huggingface.co/spaces/mosaicml/mpt-7b-instruct) for more information.
+[MosaicML’s inference service](https://www.mosaicml.com/blog/inference-launch) allows users to deploy their ML models and run inference on them. In this folder, we provide an example of how to use MPT-7B, a state-of-the-art 6.7B parameter instruction fine-tuned language model trained by MosaicML for inference. Check out the [MPT-Instruct huggingface repo](https://huggingface.co/spaces/mosaicml/mp ...
```
this text is just about instruct, but the example is about all the variants. so lets change the text to mention all the variants/not only talk about the instruct model

dakinggg commented on examples/inference-deployments/mpt/README.md:
```diff
@@ -1,22 +1,57 @@
+    - `mpt_7b.yaml` - a yaml to deploy [MPT-7B Base](https://huggingface.co/mosaicml/mpt-7b).
+    - `mpt_7b_instruct.yaml` - a yaml to deploy [MPT-7B Intstruct](https://huggingface.co/mosaicml/mpt-7b-instruct).
+    - `mpt_7b_storywriter.yaml` - a yaml to deploy [MPT-7B StoryWriter](https://huggingface.co/mosaicml/mpt-7b-storywriter).
+- Model handlers - these define how your model should be loaded and what should happen in a forward pass. You can use the default handlers here or write your custom model handler as per instructions [here](https://docs.mosaicml.com/projects/mcli/en/latest/inference/deployment_features.html#custom-model-handlers).
```
```suggestion
- Model handlers - these define how your model should be loaded and how the model should be run when receiving a request. You can use the default handlers here or write your custom model handler as per instructions [here](https://docs.mosaicml.com/projects/mcli/en/latest/inference/deployment_features.html#custom-model-handlers).
```

dakinggg commented on examples/inference-deployments/mpt/README.md:
```diff
@@ -1,22 +1,57 @@
+    - `mpt_7b_instruct.yaml` - a yaml to deploy [MPT-7B Intstruct](https://huggingface.co/mosaicml/mpt-7b-instruct).
+    - `mpt_7b_storywriter.yaml` - a yaml to deploy [MPT-7B StoryWriter](https://huggingface.co/mosaicml/mpt-7b-storywriter).
+- Model handlers - these define how your model should be loaded and what should happen in a forward pass. You can use the default handlers here or write your custom model handler as per instructions [here](https://docs.mosaicml.com/projects/mcli/en/latest/inference/deployment_features.html#custom-model-handlers).
+    - `mpt_7b_handler.py` - a python script using DeepSpeed.
```
can we be a bit more descriptive than "python script"? That description makes me think i can just run it locally

dakinggg reviewed (commented):
I suggest also improving the top level readme as part of this pr that should introduce the inference service, explain how to setup your environment, explain how each example is structured (what is a model handler, what does the yaml do, etc). Then you can link to that from this readme for the basic instructions.

dakinggg commented on examples/inference-deployments/mpt/README.md:
````diff
@@ -39,3 +74,12 @@ To deploy, simply run `mcli deploy -f mpt_7b_instruct.yaml --cluster <your_clust
 ```
+
+
+## Before you go
````
would be good to include a "whats next". we can brainstorm some standard next steps to include across each example, but basic ones would be to point them to come train with us :)

dakinggg commented on examples/inference-deployments/mpt/README.md:
```diff
@@ -1,22 +1,57 @@
+
+
+
+## Sending requests to your deployment
```
Can we also include example usage with langchain? https://github.com/hwchase17/langchain/blob/master/docs/modules/models/llms/integrations/mosaicml.ipynb

Would also be nice to make the different ways to query it expandable sections, like
"With MCLI"
"With curl"
"With langchain"
"With python"

sakshigupta93 replied:
![Screenshot 2023-06-09 at 4 34 29 PM](https://github.com/mosaicml/examples/assets/1689437/77213409-5395-4f60-9f4c-dfcac9dc2a51)
Added!

dakinggg reviewed (commented):
I think the readme should also explain how to deploy your model from cloud object store

sakshigupta93 commented:
Added deployment from cloud info as well, and updated the screenshots in the PR description!
![Screenshot 2023-06-09 at 4 55 49 PM](https://github.com/mosaicml/examples/assets/1689437/5ad6da2c-69fd-4829-b229-8e0cdb0faa34)

sakshigupta93 commented:
Addressed almost all comments except the What's Next section, which I'll add soon as well. PTAL at the current state/screenshots! At a high level if this looks good then I'll also add up the Instructor/Diffusion models READMEs!

dakinggg commented on examples/inference-deployments/README.md:
```diff
@@ -1,11 +1,36 @@
+### YAMLs
 
-Once you have `mcli` set up, take a look at the README for each of the models in this repo for the command to deploy the model.
+Deployment submissions to the MosaicML platform can be configured through a YAML file or using our Python API’s InferenceDeploymentConfig class. We have provided YAMLs in these examples which contain information like name, image, download path for the model, and more. Please see [this link](https://docs.mosaicml.com/projects/mcli/en/latest/inference/inference_schema.html) to understand what these ...
```
```suggestion
Deployment submissions to the MosaicML platform can be configured through a YAML file or using our Python API’s `InferenceDeploymentConfig` class. We have provided YAMLs in these examples which contain information like name, image, download path for the model, and more. Please see [this link](https://docs.mosaicml.com/projects/mcli/en/latest/inference/inference_schema.html) to understand what these parameters mean.
```

Are there docs for the inferencedeploymentconfig we can link to?

dakinggg commented on examples/inference-deployments/mpt/README.md:
```diff
@@ -1,34 +1,130 @@
+
+## Sending requests to your deployment
+
+Once the deployment is ready, it's time to run inference! 
```
add a line here about how to find your deployment's name?

dakinggg commented on examples/inference-deployments/mpt/README.md:
````diff
@@ -1,34 +1,130 @@
+<summary> Using Curl </summary>
+
+```bash
+curl https://<deployment-name>.inf.hosted-on.mosaicml.hosting/predict_stream \
````
should this example be with predict stream? just not sure if that was intentional or not

dakinggg commented on examples/inference-deployments/mpt/README.md:
````diff
@@ -1,34 +1,130 @@
+<summary> Using Langchain </summary>
+
+```python
+# Sign up for an account: https://forms.mosaicml.com/demo?utm_source=langchain
````
can probably delete this comment, sign up is covered earlier in the readme

dakinggg commented on examples/inference-deployments/mpt/README.md:
```diff
@@ -1,34 +1,130 @@
+| Parameters | Type | Required | Default | Description |
+| --- | --- | --- | --- | --- |
+| input_string | List[str] | yes | N/A | The prompt to generate a completion for. |
+| top_p | float | no | 0.95 | Defines the tokens that are within the sample operation of text generation. Add tokens in the sample for more probable to least probable until the sum of the probabilities is greater than top_p |
```
I think we support all HF generate kwargs here?

sakshigupta93 replied:
Nope, there's a known issue where we accidentally convert all the floats to int, so removed them from this list. This is what we support at the moment, ideally we will fix that bug sometime to be able to support the rest of the kwargs lol

dakinggg replied:
ah ok

dakinggg commented on examples/inference-deployments/README.md:
```diff
@@ -1,11 +1,36 @@
+## Content
+We have provided examples for 3 different model types in this repo:
+
+1. MPT series
```
can we make these links to the folders?

dakinggg commented on examples/inference-deployments/README.md:
```diff
@@ -1,11 +1,36 @@
+
 
-After getting started, follow the [MCLI Docs](https://docs.mosaicml.com/projects/mcli/en/latest/quick_start/quick_start_inference.html) for quick start instructions.
+### YAMLs
```
Can we add a brief section here mentioning the inference image, high level what it contains, and noting that it is pinned in all the examples?

dakinggg commented on examples/inference-deployments/mpt/README.md:
```diff
@@ -1,34 +1,131 @@
+## Inference with MPT-7B 
 
-A state-of-the-art 6.7B parameter instruction finetuned language model trained by MosaicML. Check out the [MPT-Instruct huggingface repo](https://huggingface.co/spaces/mosaicml/mpt-7b-instruct) for more information.
+[MosaicML’s inference service](https://www.mosaicml.com/blog/inference-launch) allows users to deploy their ML models and run inference on them. In this folder, we provide an example of how to use MPT-7B, a state-of-the-art 6.7B parameter instruction fine-tuned language model trained by MosaicML for inference. Check out [this link](https://www.mosaicml.com/blog/mpt-7b) for more information!
```
```suggestion
[MosaicML’s inference service](https://www.mosaicml.com/blog/inference-launch) allows users to deploy their ML models and run inference on them. In this folder, we provide an example of how to use MPT-7B, a family of 6.7B parameter large language models, including the base model, an instruction finetuned variant, and a variant finetuned on long context books. Check out [this link](https://www.mosaicml.com/blog/mpt-7b) for more information!
```

dakinggg reviewed (approved):
Looking good, thank you!

dakinggg commented:
One more suggestion. Its pretty common for users to be confused about when their deployment is actually ready and also when there are errors, suggesting adding a note about how to use `mcli get deployment logs`, both to see when its ready, and to see traceback when there is an error
